platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ResolutionHost.java

```diff
@@ -42,13 +50,71 @@ default DisplayName displayName(String type) {
     }
 
     /**
-     * Rethrows the provided failures. Does nothing if the list of failures is empty.
+     * Returns the problems service this host can use for reporting failures.
+     *
+     * @return the problems service
      */
-    default void rethrowFailure(String type, Collection<Throwable> failures) {
-        mapFailure(type, failures).ifPresent(e -> {
+    InternalProblems getProblems();
+
+    /**
+     * Rethrows the provided failures, doing nothing if the list of failures is empty.
+     * <p>
+     * If any of the failures (or their ancestor causes) are {@link ReportableAsProblem}, they will all be reported to the problems
+     * service available on this type via {@link #getProblems()}.
+     *
+     * @param resolutionType what was resolved, e.g. "dependencies", "artifacts", "files"
+     * @param failures the exceptions encountered during resolution
+     */
+    default void rethrowFailuresAndReportProblems(String resolutionType, Collection<Throwable> failures) {
+        reportProblems(failures);
+        consolidateFailures(resolutionType, failures).ifPresent(e -> {
             throw e;
         });
     }
 
-    Optional<? extends ResolveException> mapFailure(String type, Collection<Throwable> failures);
+    /**
+     * Consolidates the given failures into a single exception, if there are any.
+     *
+     * @param resolutionType what was resolved, e.g. "dependencies", "artifacts", "files"
+     * @param failures the exceptions encountered during resolution
+     * @return an {@link org.gradle.internal.exceptions.MultiCauseException} containing all the failures, or {@link Optional#empty()} if there are no failures
+     */
+    Optional<? extends ResolveException> consolidateFailures(String resolutionType, Collection<Throwable> failures);
+
+    /**
+     * If the given failure (or their ancestor causes) are {@link ReportableAsProblem}, they will all be reported to the problems
+     * service available on this type via {@link #getProblems()}.
+     *
+     * @param failure the exception to inspect
+     */
+    default void reportProblems(Throwable failure) {
+        reportProblems(Collections.singleton(failure));
+    }
+
+    /**
+     * If the given failures (or their ancestor causes) are {@link ReportableAsProblem}, they will all be reported to the problems
+     * service available on this type via {@link #getProblems()}.
+     *
+     * @param failures the exceptions to inspect
+     */
+    default void reportProblems(Collection<Throwable> failures) {
+        Queue<Throwable> exceptionQueue = new LinkedList<>(failures);
+        Throwable current = exceptionQueue.poll();
```

Comment:
❌ I think we should try to write this in a safer manner by keeping track of the `seen` exceptions -- to avoid edge cases like circular `cause` references

Reply:
We should never have circular cause exceptions though.  We'd have to go to some lengths to create that situation, since typically the cause is only supplied at exception construction.

Other places where we check exception causes don't do this, so I don't think it's worth adding here.